### PR TITLE
Add support for usernames when authenticating

### LINKED ISSUE #188: Support authentication with username and password
Redis 6 will introduce the concept of different users as part of the new ACL support.

We currently only support configuring a password via `DialAuthPass` and ignore usernames set via the URL/DSN.

I propose we...

1. Add a new function `DialAuthUser` that takes a username and a password.
2. Stop ignoring the username provided via URL/DSN.
3. Allow setting the username via URL parameter

If both a username and password is provided (either via `DialAuthUser` or via DSN/URL) radix should automatically try sending both. If only a password and no username is provided it should just send the password. This should work since Redis 6 still supports the old `AUTH` syntax (the old syntax `AUTH <password>` gets interpreted as `AUTH default <password>`).

This will break users with Redis versions < 6 that set a username other than "default" in their URL/DSN, but this shouldn't be a problem and I assume most people don't do this.

### PR DESCRIPTION
**Pull request by nussjustin**
Fixes #188

### CODE REVIEW

File: radix.go [hunk 1]
```diff
@@ -404,7 +417,18 @@ func parseRedisURL(urlStr string) (string, []DialOpt) {
 
 	var opts []DialOpt
 	q := u.Query()
-	if p, ok := u.User.Password(); ok {
+
+	n := u.User.Username()
```
Comment (mediocregopher) [inline radix.go]: This whole section could probably be cleaned up into something like:

```go
const defaultAuthUser = "default" // also used in DialAuthPass

// ...

username := defaultAuthUser
if n := u.User.UserName(); n != "" {
    username = n
} else if n := q.Get("username"); n != "" {
    username = n
}

password := q.Get("password")
if p := u.User.Password(); p != nil {
    password = p
}

if password != "" {
    opts = append(opts, DialAuthUser(username, password))
}
```

Comment (nussjustin) [reply radix.go]: PTAL

I removed the condition around the `DialAuthUser` call since `Dial` already handles the default user and empty passwords

**Comment by mediocregopher** [comment]
LGTM, just needs the tests to be fixed =/ not sure if there's really a nice way to do it, might be necessary to use a Conn stub

**Comment by nussjustin** [comment]
Ah yeah, didn't think about the syntax/argument error on Redis < 6... Will take a look at using a stub later this week

**Comment by nussjustin** [comment]
Well I changed the tests to only check the username when Redis 6 is detected, since using a Stub doesn't really work here (`Dial` always connects to an actual socket).

The tests should probably be refactored at some point in the future to actually test what is send to redis, but I can't find the motivation to do this right now.

**Comment by mediocregopher** [comment]
Lol, fair enough. Thanks!

**Pull request merged**